File: backend/src/main/kotlin/com/example/backend/controllers/ProduitController.kt
package com.example.backend.controllers

import com.example.backend.services.ProduitService
import com.example.backend.dtos.*
import com.example.backend.exceptions.NotFoundException
import com.example.backend.exceptions.ValidationException
import com.example.backend.models.RetourProduit
import com.example.backend.services.ProduitRetourRequestDto
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/produits")
class ProduitController(
  private val produitService: ProduitService
) {

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping
  fun createProduit(@RequestBody request: ProduitRequestDto): ResponseEntity<Any> {
    return try {
      ResponseEntity.status(HttpStatus.CREATED).body(produitService.createProduit(request))
    } catch (e: ValidationException) {
      ResponseEntity.badRequest().body(mapOf("error" to e.message))
    } catch (e: NotFoundException) {
      ResponseEntity.status(HttpStatus.NOT_FOUND).body(mapOf("error" to e.message))
    } catch (e: Exception) {
      ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
        .body(mapOf("error" to "Erreur interne: ${e.localizedMessage}"))
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/{id}")
  fun getProduitById(@PathVariable id: Int): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(produitService.getProduitById(id))
    } catch (e: NotFoundException) {
      ResponseEntity.status(HttpStatus.NOT_FOUND).body(mapOf("error" to e.message))
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/{id}/map")
  fun getProduitByIdMap(@PathVariable id: Int): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(produitService.getProduitByIdMap(id))
    } catch (e: NotFoundException) {
      ResponseEntity.status(HttpStatus.NOT_FOUND).body(mapOf("error" to e.message))
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/{id}/info")
  fun getProduitDetailById(@PathVariable id: Int): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(produitService.getProduitDetailById(id))
    } catch (e: NotFoundException) {
      ResponseEntity.status(HttpStatus.NOT_FOUND).body(mapOf("error" to e.message))
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/{id}/info/en_rayon")
  fun getProduitEnRayonDetailById(@PathVariable id: Int): ResponseEntity<List<Map<String,Any?>>> {
    return ResponseEntity.ok(produitService.getProduitEnRayonDetailById(id))

  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/{id}/info/scan/en_rayon")
  fun getEnRayonDetailById(@PathVariable id: String): ResponseEntity<Map<String,Any?>> {
    return ResponseEntity.ok(produitService.getEnRayonDetailById(id))
  }

//  @CrossOrigin(origins = ["http://localhost:4200"])
//  @PreAuthorize("isAuthenticated()")
//  @GetMapping("/{id}/info/scan/en_rayon/param")
//  fun searchProductsParam(
//    @PathVariable id: String,
//    @RequestParam(required = false) rayonId: String?,
//    @RequestParam(required = false) fabriquantId: String?,
//    @RequestParam(required = false) etagereId: String?,
//    @RequestParam(required = false) formeId: String?,
//    @RequestParam(required = false) magasinId: String?,
//    @RequestParam(required = false) categorieId: String?
//  ): ResponseEntity<Map<String,Any?>> {
//    val result = produitService.getEnRayonDetailByIdParam(id, rayonId!!.toInt(), fabriquantId!!.toInt(),etagereId!!.toInt(),formeId!!.toInt(),magasinId!!.toInt(),categorieId!!.toInt())
//    return ResponseEntity.ok(result)
//  }

  @PreAuthorize("isAuthenticated()")
  @GetMapping
  fun getAllProduits(
    @RequestParam(defaultValue = "0") page: String,
    @RequestParam(defaultValue = "10") size: String,
    @RequestParam(required = false) search: String,
    @RequestParam(defaultValue = "id") sortBy: String
  ): ResponseEntity<Page<ProduitResponseDto>> {
    val pageNumber = page.toIntOrNull() ?: 0 // Default to 0 if conversion fails
    val pageSize = size.toIntOrNull() ?: 10 // Default to 10 if conversion fails
    val produits = produitService.getAllProduits(PageRequest.of(pageNumber, pageSize, Sort.by(sortBy)))
    return ResponseEntity.ok(produits)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/products/search")
  fun searchProducts(
    @RequestParam("query") query: String,
    @RequestParam("page", defaultValue = "0") page: Int,
    @RequestParam("size", defaultValue = "10") size: Int,
  ): ResponseEntity<Page<Map<String,Any?>>> {
    val result = produitService.searchProducts(query, page, size)
    return ResponseEntity.ok(result)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/products/search/param")
  fun searchProductsParam(
    @RequestParam("query") query: String,
    @RequestParam("page", defaultValue = "0") page: Int,
    @RequestParam("size", defaultValue = "10") size: Int,
    @RequestParam(required = false) rayonId: String?,
    @RequestParam(required = false) fabriquantId: String?,
    @RequestParam(required = false) etagereId: String?,
    @RequestParam(required = false) formeId: String?,
    @RequestParam(required = false) magasinId: String?,
    @RequestParam(required = false) categorieId: String?
  ): ResponseEntity<Page<Map<String,Any?>>> {
    val result = produitService.searchProductsWithParam(query, page, size,rayonId, fabriquantId, etagereId, formeId, magasinId, categorieId)
    return ResponseEntity.ok(result)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PutMapping("/{id}")
  fun updateProduit(@PathVariable id: Int, @RequestBody request: ProduitRequestDto): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(produitService.updateProduit(id, request))
    } catch (e: ValidationException) {
      ResponseEntity.badRequest().body(mapOf("error" to e.message))
    } catch (e: NotFoundException) {
      ResponseEntity.status(HttpStatus.NOT_FOUND).body(mapOf("error" to e.message))
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @DeleteMapping("/{id}")
  fun deleteProduit(@PathVariable id: Int): ResponseEntity<Any> {
    return try {
      produitService.deleteProduit(id)
      ResponseEntity.noContent().build()
    } catch (e: NotFoundException) {
      ResponseEntity.status(HttpStatus.NOT_FOUND).body(mapOf("error" to e.message))
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/stock")
  fun updateStockProduit(
    @PathVariable id: Int,
    @RequestBody request: ProduitStockUpdateRequestDto
  ): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(produitService.updateStockProduit(id, request))
    } catch (e: ValidationException) {
      ResponseEntity.badRequest().body(mapOf("error" to e.message))
    } catch (e: NotFoundException) {
      ResponseEntity.status(HttpStatus.NOT_FOUND).body(mapOf("error" to e.message))
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/tarification")
  fun updateTarificationProduit(
    @PathVariable id: Int,
    @RequestBody request: ProduitTarificationUpdateRequestDto
  ): ResponseEntity<Any> {
    return try {
      ResponseEntity.ok(produitService.updateTarificationProduit(id, request))
    } catch (e: NotFoundException) {
      ResponseEntity.status(HttpStatus.NOT_FOUND).body(mapOf("error" to e.message))
    }
  }

  // Endpoints for Categorie, Fournisseur, Depot, Rayon
  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/categories")
  fun createCategorie(@RequestBody dto: CategorieDto) =
    ResponseEntity.status(HttpStatus.CREATED).body(produitService.createCategorie(dto))

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/categories")
  fun getAllCategories() = ResponseEntity.ok(produitService.getAllCategories())

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/fournisseurs")
  fun createFournisseur(@RequestBody dto: FournisseurDto) =
    ResponseEntity.status(HttpStatus.CREATED).body(produitService.createFournisseur(dto))

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/fournisseurs")
  fun getAllFournisseurs() = ResponseEntity.ok(produitService.getAllFournisseurs())

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/rayons")
  fun createRayon(@RequestBody dto: RayonDto) =
    ResponseEntity.status(HttpStatus.CREATED).body(produitService.createRayon(dto))

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/rayons")
  fun getAllRayons() = ResponseEntity.ok(produitService.getAllRayons())


  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/{id}/details")
  fun getProduitDetails(@PathVariable id: Int): ResponseEntity<Map<String, Any?>> {
    return ResponseEntity.ok(produitService.getProduitDetails(id))
  }

}

File: backend_py/pharmaPython/app/api/v1/produit_controller.py
from fastapi import APIRouter, Depends, HTTPException, Query, Path, Body
from sqlalchemy.orm import Session
from typing import Any, Dict, List, Optional

from app.api.deps import get_db
from app.schemas.produit_detail_dto import ProduitRequestDto, ProduitResponseNewDto, \
  ProduitTarificationUpdateRequestDto, CategorieDto, FournisseurDto, RayonDto, ProduitRequestNewDto, \
  ProduitStockUpdateRequestDto
from app.services.produit_service import ProduitService

router = APIRouter(prefix="/api/produits", tags=["Produits"])

@router.post("/", status_code=201)
def create_produit(request: ProduitRequestDto, db: Session = Depends(get_db)):
  try:
    return ProduitService(db, ...).create_produit(request)
  except ValueError as e:
    raise HTTPException(status_code=400, detail=str(e))

@router.post("/new", status_code=201)
def create_produit_new(request: ProduitResponseNewDto, db: Session = Depends(get_db)):
  return ProduitService(db, ...).create_produit_new(request)

@router.get("/{id}")
def get_produit_by_id(id: int, db: Session = Depends(get_db)):
  return ProduitService(db, ...).get_produit_by_id(id)

@router.get("/{id}/map")
def get_produit_by_id_map(id: int, db: Session = Depends(get_db)):
  return ProduitService(db, ...).get_produit_by_id_map(id)

@router.get("/{id}/info")
def get_produit_detail_by_id(id: int, db: Session = Depends(get_db)):
  return ProduitService(db, ...).get_produit_detail_by_id(id)

@router.get("/{id}/info/en_rayon")
def get_produit_enrayon_detail_by_id(id: int, db: Session = Depends(get_db)):
  return ProduitService(db, ...).get_produit_enrayon_detail_by_id(id)

@router.get("/{id}/info/scan/en_rayon")
def get_enrayon_detail_by_id(id: str, db: Session = Depends(get_db)):
  return ProduitService(db, ...).get_enrayon_detail_by_id(id)

@router.get("/")
def get_all_produits(
  page: int = Query(0, ge=0),
  size: int = Query(10, ge=1),
  sortBy: str = Query("id"),
  db: Session = Depends(get_db)
):
  return ProduitService(db, ...).get_all_produits(page, size)

@router.get("/products/search")
def search_products(query: str, page: int = 0, size: int = 10, db: Session = Depends(get_db)):
  return ProduitService(db, ...).search_products(query, page, size)

@router.get("/products/search/param")
def search_products_param(
  query: Optional[str] = None, page: int = 0, size: int = 10,
  rayonId: Optional[str] = None, fabriquantId: Optional[str] = None,
  etagereId: Optional[str] = None, formeId: Optional[str] = None,
  magasinId: Optional[str] = None, categorieId: Optional[str] = None,
  db: Session = Depends(get_db)
):
  return ProduitService(db, ...).search_products_with_param(
    query=query, page=page, size=size,
    rayonId=rayonId, fabriquantId=fabriquantId, etagereId=etagereId,
    formeId=formeId, magasinId=magasinId, categorieId=categorieId
  )

@router.put("/{id}")
def update_produit(id: int, request: ProduitRequestDto, db: Session = Depends(get_db)):
  return ProduitService(db, ...).update_produit(id, request)

@router.put("/{id}/save")
def add_or_update_produit_new(id: int, request: ProduitRequestNewDto, db: Session = Depends(get_db)):
  return ProduitService(db, ...).add_or_update_produit_new(id, request)

@router.delete("/{id}", status_code=204)
def delete_produit(id: int, db: Session = Depends(get_db)):
  ProduitService(db, ...).delete_produit(id)
  return {"message": "deleted"}

@router.post("/{id}/stock")
def update_stock_produit(id: int, request: ProduitStockUpdateRequestDto, db: Session = Depends(get_db)):
  return ProduitService(db, ...).update_stock_produit(id, request)

@router.post("/{id}/tarification")
def update_tarification_produit(id: int, request: ProduitTarificationUpdateRequestDto, db: Session = Depends(get_db)):
  return ProduitService(db, ...).update_tarification_produit(id, request)

@router.post("/categories", status_code=201)
def create_categorie(dto: CategorieDto, db: Session = Depends(get_db)):
  return ProduitService(db, ...).create_categorie(dto)

@router.get("/categories")
def get_all_categories(db: Session = Depends(get_db)):
  return ProduitService(db, ...).get_all_categories()

@router.post("/fournisseurs", status_code=201)
def create_fournisseur(dto: FournisseurDto, db: Session = Depends(get_db)):
  return ProduitService(db, ...).create_fournisseur(dto)

@router.get("/fournisseurs")
def get_all_fournisseurs(db: Session = Depends(get_db)):
  return ProduitService(db, ...).get_all_fournisseurs()

@router.post("/rayons", status_code=201)
def create_rayon(dto: RayonDto, db: Session = Depends(get_db)):
  return ProduitService(db, ...).create_rayon(dto)

@router.get("/rayons")
def get_all_rayons(db: Session = Depends(get_db)):
  return ProduitService(db, ...).get_all_rayons()

@router.get("/{id}/details")
def get_produit_details(id: int, db: Session = Depends(get_db)):
  return ProduitService(db, ...).get_produit_details(id)
